c-scape/src/lib.rs

```diff
@@ -2423,10 +2424,69 @@ unsafe extern "C" fn nanosleep(
 
 // exec
 
+unsafe fn null_terminated_array<'a>(list: *const *const c_char) -> Vec<&'a CStr> {
+    let mut len = 0;
+    while !(*list.wrapping_add(len)).is_null() {
+        len += 1;
+    }
+    let list = std::slice::from_raw_parts(list, len);
+    list.into_iter()
+        .copied()
+        .map(|ptr| CStr::from_ptr(ptr))
+        .collect()
+}
+
+fn file_exists(cwd: &rsix::io_lifetimes::BorrowedFd<'_>, path: &CStr) -> bool {
+    rsix::fs::accessat(
+        cwd,
+        path,
+        rsix::fs::Access::EXISTS | rsix::fs::Access::EXEC_OK,
+        rsix::fs::AtFlags::empty(),
+    )
+    .is_ok()
+}
+
+fn resolve_binary<'a>(file: &'a CStr, envs: &[&CStr]) -> rsix::io::Result<Cow<'a, CStr>> {
+    let file_bytes = file.to_bytes();
+    if file_bytes.contains(&b'/') {
+        Ok(Cow::Borrowed(file))
+    } else {
+        let cwd = rsix::fs::cwd();
+        match envs
+            .into_iter()
+            .copied()
+            .map(CStr::to_bytes)
+            .find(|env| env.starts_with(b"PATH="))
+            .map(|env| env.split_at(b"PATH=".len()).1)
+            .unwrap_or(b".")
```

Comment:
POSIX isn't clear about what the default path should be, but on Linux systems it's typically something like "/bin:/usr/bin". It shouldn't include "." because it turns out that can actually be a security concern if the user ever happens to cd to a directory controlled by an adversary.